### ISSUE TITLE
Expected behavior for Go To Definition for metadata references

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#24349]
The Navigate to Decompiled Sources feature recently added is the first step towards a better <kbd>F12</kbd> experience. This issue describes the more complete picture for how this experience should eventually look.

🔗 [AB#624702](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/624702)

⚠️ This workflow is currently just a draft

1. Start with the metadata reference
2. Locate the correct runtime assembly from the metadata reference (accounting for execution environment and assembly binding redirection)
3. Locate the symbols for the runtime assembly
4. Using the symbol file and/or symbol server, locate the original source file(s) for the symbol getting decompiled
5. If step 3 or 4 fails, decompile the symbol from the runtime assembly using ILSpy
6. If step 2 fails, attempt steps 3-5 using the metadata reference instead of the runtime assembly
7. Otherwise, don't attempt to decompile

**Comment by tmat** [comment]
A couple of things
- I think we should first determine if we can find the original source locally - in the PDB (embedded), in debugger source cache, etc. If we can we should display it.
- Otherwise, we should use the current Metadata As Source implementation that only displays signatures and doc comments. The benefit is that you don't need to agree to anything legal at that point and it's fast. Based on my personal experience in most cases I just want to see the signatures of the public APIs and doc comments. 
- Then we should have 2 gestures (like expanding a collapsed method body or buttons in gold bar etc.):
   1) One that triggers symbol and source resolution from symbol servers. 
   2) One that runs disassembly.

**Comment by sharwell** [comment]
> I think we should first determine if we can find the original source locally - in the PDB (embedded), in debugger source cache, etc.

This is step 4 of the original workflow.

> Otherwise, we should use the current Metadata As Source implementation that only displays signatures and doc comments. The benefit is that you don't need to agree to anything legal at that point and it's fast.

This is a somewhat orthogonal discussion (if/how to trigger Go To Definition separately from "Show Signatures"). Given the decompiler is currently only enabled at the explicit request of a developer, I believe it's reasonable to assume that the use of ILSpy in these cases is the desired behavior.

**Comment by athinboy** [comment]
Is true! Is true:

I support this:

https://github.com/dotnet/sourcelink/issues/169



@ctaggart Yes, we need a tool that navigates directly to the source in the IDE so that we can analyze the underlying implementation and debug the current code directly at run time.
@chuckries Java in the IDE to download the source code, can be directly in the current project to view the source code, such as the JDK source code, observe these implementations, can improve their development capabilities, and. NET to view the source code is complex, we need to use the Anti-compilation tool (Ilspy) or from git to get the source code

I think it has a positive effect on the development of the community and can benefit more. NET Developers


https://github.com/dotnet/sourcelink

**Comment by athinboy** [comment]
I think the people of dotnet foundation are "stand still and refuse to make progress ", they don't and wouldn't to think about why more and more .net programmer turn to Java!

**Comment by oliverjanik** [comment]
Is there any progress on this? Resharper has had this for a decade now.

This is extremely useful functionality, for quickly verifying assumptions instead of scouring the internet for source files and documentation!

**Comment by jogibear9988** [comment]
isnt dnspy the right tool for the debugging? https://github.com/0xd4d/dnSpy

**Comment by gautelo** [comment]
@oliverjanik I agree. Visual Studio is getting to a point where I almost don't need resharper anymore. With [IntelisenseExtender](https://marketplace.visualstudio.com/items?itemName=Dreamescaper.IntelliSenseExtender) (which btw, should also be a standard option), this is basically the only thing I really need to be able to move to a leaner non-resharper world.

When you've split your business code across several nuget packages with interdependencies, this sort of support is a pretty vital analysis feature, so I hope this will get some more attention soon.

Wish I could help, but I'm way too scared of anything this close to metal. ;)

**Comment by Zagrophyte** [comment]
Honestly, even having just the metadata available would be a massive step up for me. Often, I'm just looking for what members a given library type or class supports. 

I notice the hover tooltip over a type in VS already shows a truncated form of this information--maybe this could take a multi-phased approach?

Currently even the metadata is not available in its full form for display via Go To Definition for external libraries.

**Comment by Dargazo** [comment]
Honestly, with .NET Core being open sourced, I kind of expected to be able to F12 into .NET Core functions. My immediate reaction to the current F12 behavior for .NET Core was "well where's the source code?" (now I know I can find it on github, but still...). So for me, this issue feels like a logical feature following from the awesome open sourcing work done by the .NET Core team. I would really like this feature. :)

**Comment by jnm2** [comment]
Embedded source and SourceLinked source should be preferred over decompiled source. This is one of the things that makes me turn ReSharper back on.

Oren Novotny, Kevin Pilch, and Rich Lander suggested this too: https://twitter.com/runfaster2000/status/1148348176283234304 so maybe we'll see something soon?

**Comment by anangaur** [comment]
Related ask from: https://developercommunity.visualstudio.com/content/idea/468799/integrated-decompiler-for-c-or-other-supported-lan.html

visual studio should have a de-compiler for nuget packages or an option to view source code when clicked on a method or class, and other options like go to implementation scope is limited to project code but it should be extended to nuget package source also for now it only shows class metadata but full source must be shown

there are other plugins like dotpeek but paid not all can offord that

**Comment by alefranz** [comment]
What are the work items on roslyn and sourcelink needed to make this happen?
Has someone investigated it already?
If there is a high level design it would be nice to flesh it out so the community can help picking up the work.
I would love to help on this but I expect this requires some work on the Visual Studio side, so I believe it can not be implemented without some major upfront design.

Thank you!

**Comment by mikadumont** [comment]
Customer Feedback: 

"Setting a breakpoint in source should work, even if the source was decompiled.  If from Sourcelink, it works, but you’re left in this weird not-quite-real-source mode where you can’t F12 on many types.  It seems like VS loses all references and is in some disconnected source.  If in decompiled source, I need symbols to also be generated and imported."

**Comment by MeikTranel** [comment]
@mikadumont i feel like this moves the goalpost a little to much. 

This issue is about the current state of SourceLink clients forcing users to do backflips to something technically simple. The customer feedback you posted is coming from a place where we would need to be able to:
- Write out symbols from decompiled IL with line information picked up by the debugger
- magic ahead-of-time download of source files (whereas right now sourcelink downloads source files on-demand)

This is a massive increase in effort in comparison to what this issue actually only asks for:  Execute the same command on Go-To-Definition that is executed when F11'ing into sourcelinked assemblies.

I agree with the customers sentiment - it was what all this at the end should look like. But it goes way beyond this issue, which would only prolong this issue's lifetime if we were to add it to the goals of this issue.

**Comment by oliverjanik** [comment]
Hello again, we're nearing almost 2 years since this issue was first created. 

Now that we have 3.1 out of the door (big thanks for lots of amazing features), has there been any progress on this issue?

Seems like it's assigned Priority 1, which makes me very hopeful for v16.5 !

**Comment by athinboy** [comment]
#https://github.com/dotnet/roslyn/issues/42525

**Comment by olexandrvynnychenko** [comment]
The same behavior in VS 16.5.4

**Comment by athinboy** [comment]
Still there is not any progress ?!

**Comment by baratgabor** [comment]
I can't add anything useful, besides remarking that this feature would be super-useful for ASP.NET Core, because it uses a myriad of extension methods that configure and inject services into the DI container, and it's often necessary to know what is exactly happening to avoid misconfiguration, incorrect ordering or redundant code.

**Comment by athinboy** [comment]
Come on ! Three years !
On this feature ,Java is far ahead!
Does mircosoft abandon dotnet ?!

**Comment by s-bauer** [comment]
@athinboy I think there have been a few changes in some of the latest preview versions of Visual Studio, but don’t quote me in that.

**Comment by sharwell** [comment]
@athinboy This is a priority work item, but it's taking time to ensure the feature works seamlessly across all the different ways users can interact with source code. In the meantime, the Navigate to Decompiled Sources feature can be enabled to provide a portion of this experience. 😄

**Comment by athinboy** [comment]
@sharwell 
thans for your work!
I mean,it progress slow and late. there are too many software that without  complete、exact  document，even the source code of  roslyn、of dotnet。
for example, in https://www.cnblogs.com/,high qualified documents  become  fewer and fewer ，because the people can not 
to check and debug the source code .
i think it is top 1 priority that to browse 、check、debug the source code conveniently, it is proved in java.
sorry for my complaint and bad english. but what i said is true!

**Comment by xyz0822** [comment]
@sharwell it seems blocked on partner team,what does it mean blocked? is there any process on this?

**Comment by WeihanLi** [comment]
Really hope for this feature, big thanks for the team working on this.

**Comment by mxashlynn** [comment]
I know everyone is hard at work on this, I just wanted to say that it is still something I'd love to see eventually!

**Comment by maloo** [comment]
@mikadumont , does the "Dev17 IDE Prio" mean it will not be done in VS2019? We have to wait for VS2022 to navigate to source code?

**Comment by DillonN** [comment]
It doesn't seem to work fully as expected in VS2022 preview 4.1 for me. It works fine for decompiling sources from NuGet libraries, but F12ing into anything from the runtime/aspnetcore you just see metadata files. I.e. just declarations and xmldocs, all the implementations are `throw null;`

**Comment by BearsPunch** [comment]
> It doesn't seem to work fully as expected in VS2022 preview 4.1 for me. It works fine for decompiling sources from NuGet libraries, but F12ing into anything from the runtime/aspnetcore you just see metadata files. I.e. just declarations and xmldocs, all the implementations are `throw null;`

Why do you think that this feature was already added to VS? As I see the ticket is still open

**Comment by gautelo** [comment]
👀 "In Progress" *hope surges* 👏

**Comment by DillonN** [comment]
> Why do you think that this feature was already added to VS? As I see the ticket is still open

Sorry @BearsPunch, I think I got linked here from a support ticket saying the issue was resolved already and confused myself. Excited to see it in progress!

**Comment by oliverjanik** [comment]
Has this gotten worse in VS2022 ? When I hit F12 (Go to Definition) on a framework method all I get is metadata:

![image](https://user-images.githubusercontent.com/2280140/147800089-8b3f6a11-617c-4d18-9b6f-55b4266bc29c.png)

**Comment by ProKn1fe** [comment]
> Has this gotten worse in VS2022 ? When I hit F12 (Go to Definition) on a framework method all I get is metadata:
> 
> ![image](https://user-images.githubusercontent.com/2280140/147800089-8b3f6a11-617c-4d18-9b6f-55b4266bc29c.png)

Do you have this option enabled?
![image](https://user-images.githubusercontent.com/54151358/147813440-98ce0501-213b-41dd-ab65-5a2d17065599.png)

**Comment by oliverjanik** [comment]
@ProKn1fe Yes I do, that's the first thing I checked.

**Comment by hammypants** [comment]
> Has this gotten worse in VS2022 ? When I hit F12 (Go to Definition) on a framework method all I get is metadata:
> 
> ![image](https://user-images.githubusercontent.com/2280140/147800089-8b3f6a11-617c-4d18-9b6f-55b4266bc29c.png)

+1ing this. This is the behavior I am seeing in VS2022 and it's miserable.

**Comment by zhyy2008z** [comment]
Please update the version of ilspy integrated in visual studio, it is too old!

**Comment by davidwengier** [comment]
> Please update the version of ilspy integrated in visual studio, it is too old!

The version of ILSpy used by Roslyn was updated a few weeks ago: https://github.com/dotnet/roslyn/pull/53304

**Comment by KieranDevvs** [comment]
> > Please update the version of ilspy integrated in visual studio, it is too old!
> 
> The version of ILSpy used by Roslyn was updated a few weeks ago: #53304

So in layman's terms, what's the status of this? When should we expect to be able to navigate to sources in VS?

**Comment by RikkiGibson** [comment]
The "Milestone" on the completed PR indicates what VS release it will arrive in. It looks like VS 17.2 preview 1 will have the new version of ILSpy.

**Comment by davidwengier** [comment]
> When should we expect to be able to navigate to sources in VS?

In Visual Studio 2022 v17.1 Preview you can navigate to source from embedded source, original source on disk, or assemblies that use Source Link. If any of those fail, or if you're using Visual Studio 2022 v17.0 (the stable version) then decompilation will be used to show the best source we can come up. All of that is of course dependent on various settings, type of DLL, PDB availability etc.

> So in layman's terms, what's the status of this?

As far as the items in the original description of this issue specifically, number 2 is still pending. If you're interested in what has been done and what is pending, in non-layman's terms, I'm working mainly from https://github.com/dotnet/roslyn/issues/55834

**Comment by dominikjeske** [comment]
I'm not sure if I can change configuration in VS but currently I'm using VS 2022 17.1 and 17.2 preview 1 and on both I have same experience. For example I'm trying to check source for JsonConfigurationExtensions->AddJsonFile (from Microsoft.Extensions.Configuration.Json). When using 'Go to definition' I have 'throw null' and 'Go to Implementation' gives me 'The symbol has no implementations'. I have symbol servers enabled.

![image](https://user-images.githubusercontent.com/3872348/154682957-f2f41e32-c7a6-412e-876d-6ac83b5e3269.png)

 Is there something I can configure in VS to make it work or is it still a bug?

**Comment by kirsan31** [comment]
The way it's working now (17.2.5) is too bad and not intuitive :(
For example:
```cs
private System.Timers.Timer _timer = new System.Timers.Timer(1000);

private void _timer_Tick(object sender, EventArgs e)
{
    _timer.Interval = 5000;
}
```
- _Go to Definition_ on `System.Timers.Timer` will show you meta data and then suggest to go to source. It`s ok.
![image](https://user-images.githubusercontent.com/17767561/175269519-10e5e88c-3db6-4684-8501-e229aa8d2a23.png)
-  _Go to Definition_ on `_timer.Interval` will show you meta and no source suggestion. This is bad 😥
- _Go to Implementation_ will show you meta only in any places :(

**Comment by KieranDevvs** [comment]
> The way it's working for now (VS 17.2.5 .Net 6.0) is too bad and not intuitive :( For example:
> 
> ```cs
> private System.Timers.Timer _timer = new System.Timers.Timer(1000);
> 
> private void _timer_Tick(object sender, EventArgs e)
> {
>     _timer.Interval = 5000;
> }
> ```
> 
> * _Go to Definition_ on `System.Timers.Timer` will show you meta data and then suggest to go to source. It`s ok.
>   ![image](https://user-images.githubusercontent.com/17767561/175269519-10e5e88c-3db6-4684-8501-e229aa8d2a23.png)
> * _Go to Definition_ on `_timer.Interval` will show you meta and no source suggestion. This is bad 😥
> * _Go to Implementation_ will show you meta only in any places :(

AFAIK base class library sources will only show from 17.3 and onwards because the feature to find the correct implementation assembly isn't released yet. 

(also, use `PeriodicTimer` over all the other timers.)

**Comment by davidwengier** [comment]
> and then suggest to go to source

The functionality in your screenshot is not a part of Roslyn as far as I'm aware. I suspect you have an extension that is providing that, so inconsistent behaviour might be an issue to report to the extension author.

**Comment by BearsPunch** [comment]
17.3 preview 2 added some new features about go to definition, but it still doesn’t work with .net source and libraries like entity framework

**Comment by davidwengier** [comment]
17.3 includes some more smarts for finding source for .NET 6, which should generally work if you're targetting the most recent SDK that you have installed. eg, if you are targetting .NET 6 and have a preview of .NET 7 installed, then it won't work, but if you have only .NET 6 installed, it should.

>  As I know the issue here because sdk comes without source code

Generally speaking, this is correct. Most often when you are "referencing .NET 6" you are referencing nuget packages that contain only empty reference assemblies. At runtime there are various mechanisms in play that resolve those references to real implementation DLLs, based on what you're targetting, have installed, is in your global.json, etc. etc. Roslyn does not have those same mechanisms sadly.

Hopefully over time, however, more scenarios can be supported, and more people will see success.

**Comment by BearsPunch** [comment]
Hey @davidwengier . I'm testing VS 17.3 Preview 6 and can't understand why I can go to a source (via source link) if press F12 to Select method from Queryable class or  to WriteLine method from class Console (classes and methods from default SDK), but if try to go to a method Run from class Task (or just class Task), VS navigates me to the metadata. After that I see on top yellow bar with text "An external source file for this definition is available. Go to File" and if I press the link, it opens me  System.Runtime.cs file with a code like public static `System.Threading.Tasks.Task Run(System.Action action) { throw null; }` . Don't understand why some default methods and classes can be found, but some not :)

**Comment by davidwengier** [comment]
Thanks for your comment @BearsPunch. If you see any other specific cases where navigation doesn't work it would be good to log them as new issues in future, and ideally include the output from the "Navigate to External Sources" output window category if possible, so we can better track individual bugs.

> if try to go to a method Run from class Task (or just class Task), VS navigates me to the metadata

This is fixed in 17.4 Preview 1 with https://github.com/dotnet/roslyn/pull/62406

> After that I see on top yellow bar with text "An external source file for this definition is available. Go to File"

I think that yellow bar is coming from an extension, it is not part of Roslyn as far as I know.

**Comment by mattzink** [comment]
17.4? Is that expected to come out this year? I guess we've been waiting 5 years on this issue so what's one more 😆

**Comment by BearsPunch** [comment]
> > After that I see on top yellow bar with text "An external source file for this definition is available. Go to File"
> 
> I think that yellow bar is coming from an extension, it is not part of Roslyn as far as I know.
No, I have totally virgin VS but enabled some preview features. No additional extensions for VS :)

**Comment by kirsan31** [comment]
> No, I have totally virgin VS but enabled some preview features. No additional extensions for VS :)

I have several addons installed, but they don't have such functions declared. And I also see this message:
https://github.com/dotnet/roslyn/issues/24349#issuecomment-1164193063

**Comment by BearsPunch** [comment]
> > No, I have totally virgin VS but enabled some preview features. No additional extensions for VS :)
> 
> I have several addons installed, but they don't have such functions declared. And I also see this message: [#24349 (comment)](https://github.com/dotnet/roslyn/issues/24349#issuecomment-1164193063)

oh yeah, exactly this bar I have. Maybe you have to enable some features in debug options or smth similar.  I know, this is unprofessional, but I just tried to enable everything what could be connected with source and source link and can't tell you what exactly I enabled :(

**Comment by davidwengier** [comment]
Thanks both, I found the VS feature that adds that bar, and its "Rich Code Navigation", in Preview Features. I did not have it on, but turning it on adds that bar.

<img width="308" alt="image" src="https://user-images.githubusercontent.com/754264/182950808-6df83429-e626-4c83-91c8-e2c40bdb6a82.png">

If you have feedback about that experience it would be best to log an issue through VS feedback so that the team responsible can see it.

**Comment by kirsan31** [comment]
VS **17.3.2** still not usable, more of it - it's become worse :(
got to definition on `Enum.HasFlag` will open this file (with cursor at beginning of it):
![image](https://user-images.githubusercontent.com/17767561/186612600-74333f1d-b98b-4df3-97b3-4dde6bc90cee.png)
🙄

After pressing `An external source file for this definition is available. Go to File` we will see:
![image](https://user-images.githubusercontent.com/17767561/186613589-69b93dac-2c15-45e5-846f-e128b4eae9da.png)
🤷‍♂️

**Comment by davidwengier** [comment]
That navigation issue has been fixed by https://github.com/dotnet/roslyn/pull/62406, which is in Visual Studio Preview if you want to try that, and will come to the stable release in time.

**Comment by zippy72** [comment]
Is there any way to turn off all decompilation? Personally I find the metadata is usually what I'm looking for

**Comment by CyrusNajmabadi** [comment]
Yes, disable the first two of these:

![image](https://user-images.githubusercontent.com/4564579/193910112-627abfb1-ae93-4443-a51c-7b82d190580a.png)

**Comment by CyrusNajmabadi** [comment]
@sharwell with decompilatino and source-link, do we need anything else here?

**Comment by TFTomSun** [comment]
@CyrusNajmabadi 
Is there already a possibility to authenticate in an enterprise network to resolve the source link sources?

**Comment by davidwengier** [comment]
> Is there already a possibility to authenticate in an enterprise network to resolve the source link sources? 

That would be a question that hopefully @chuckries can answer.

**Comment by chuckries** [comment]
@TFTomSun It depends. 
1. There is not currently an interactive auth mechanism for F12. You could possibly authenticate through the [External Sources Node](https://devblogs.microsoft.com/visualstudio/debugging-external-sources-with-visual-studio/#:~:text=External%20Sources%20node%20in%20Solution%20Explorer%20The%20feature,node.%20In%20debug%20mode%20%E2%80%93%20collapsed%20state%20%28default%29)
2. What exactly do you mean by "enterprise network"? Where are your repositories hosted?

**Comment by TFTomSun** [comment]
@chuckries The sources are hosted on an on premise gitlab instance. To reach that instance via https you need to authorize via 2FA. In the past Visual Studio had a built in browser window. Some people with on premise gitlab instances reported that it was possible to do the 2FA there to make source link work. I never tried it. The browser window was removed with VS2022.

Anyways, you would always have the same problem, if you are working with private repositories, even on public git instances. There should be a solution to reach the private source files, for people who are authorized.

**Comment by TFTomSun** [comment]
@chuckries Can you explain a bit more how I could authenticate for such scenarios via an external source node?

**Comment by cjmld5** [comment]
for about four years，it seems things still not resolved on VS17.7.2
for example :
0：turn on the enable navigation to source link or decompile
1：create a new project with net 6.0
2：F12 to view the net6.0 source code
![image](https://github.com/dotnet/roslyn/assets/6902006/ed4ca90c-e549-4bd7-8e8e-ce27d4bc2fea)

it seems source link and inner ilspy decompile dont works as well，all we got is the metadata not the source code
for such a mount of time，should we could think make it eaiser and reduce the complex：
remove the view to metadata，and just decompile the source code through F12 and make it works on edit mode and debugger mode?

on the other side:jetbrains resharper is really good，they really thinks about the developers

**Comment by zhyy2008z** [comment]
> for about four years，it seems things still not resolved on VS17.7.2 for example : 0：turn on the enable navigation to source link or decompile 1：create a new project with net 6.0 2：F12 to view the net6.0 source code ![image](https://user-images.githubusercontent.com/6902006/265940260-ed4ca90c-e549-4bd7-8e8e-ce27d4bc2fea.png)
> 
> it seems source link and inner ilspy decompile dont works as well，all we got is the metadata not the source code for such a mount of time，should we could think make it eaiser and reduce the complex： remove the view to metadata，and just decompile the source code through F12 and make it works on edit mode and debugger mode?
> 
> on the other side:jetbrains resharper is really good，they really thinks about the developers

It may be due to GFW, try jailbreaking it. I can see the source link normally work here.
![image](https://github.com/dotnet/roslyn/assets/5698463/3fe7cd86-2423-4d2e-a261-d9e41cf91fc7)

**Comment by cjmld5** [comment]
> > for about four years，it seems things still not resolved on VS17.7.2 for example : 0：turn on the enable navigation to source link or decompile 1：create a new project with net 6.0 2：F12 to view the net6.0 source code ![image](https://user-images.githubusercontent.com/6902006/265940260-ed4ca90c-e549-4bd7-8e8e-ce27d4bc2fea.png)
> > it seems source link and inner ilspy decompile dont works as well，all we got is the metadata not the source code for such a mount of time，should we could think make it eaiser and reduce the complex： remove the view to metadata，and just decompile the source code through F12 and make it works on edit mode and debugger mode?
> > on the other side:jetbrains resharper is really good，they really thinks about the developers
> 
> It may be due to GFW, try jailbreaking it. I can see the source link normally work here. ![image](https://user-images.githubusercontent.com/5698463/266371021-3fe7cd86-2423-4d2e-a261-d9e41cf91fc7.png)

@zhyy2008z if due to the GFW，i mean，the inner ILSpy should also works as well，but it seems not work and fallback to the view of metadata

**Comment by davidwengier** [comment]
Whats GFW?

The reason for the lack of real source in your situation is likely due to targeting .NET 6, but using a later SDK. In that case your projects are probably referencing nuget packages that literally do not contain any code, or SourceLink information, so there is nothing that Roslyn can use to find the right code. At runtime you'll roll forward and use whatever SDK is installed based on various rules and logic in the SDK resolvers, but Roslyn does not have that logic or information. We made it a little smarter, and it has some logic to pull the real info from SDK targetting packs, and nuget packages that contain both reference and implementation assemblies, but it doesn't cover all scenarios.

**Comment by zhyy2008z** [comment]
> > > for about four years，it seems things still not resolved on VS17.7.2 for example : 0：turn on the enable navigation to source link or decompile 1：create a new project with net 6.0 2：F12 to view the net6.0 source code ![image](https://user-images.githubusercontent.com/6902006/265940260-ed4ca90c-e549-4bd7-8e8e-ce27d4bc2fea.png)
> > > it seems source link and inner ilspy decompile dont works as well，all we got is the metadata not the source code for such a mount of time，should we could think make it eaiser and reduce the complex： remove the view to metadata，and just decompile the source code through F12 and make it works on edit mode and debugger mode?
> > > on the other side:jetbrains resharper is really good，they really thinks about the developers
> > 
> > 
> > It may be due to GFW, try jailbreaking it. I can see the source link normally work here. ![image](https://user-images.githubusercontent.com/5698463/266371021-3fe7cd86-2423-4d2e-a261-d9e41cf91fc7.png)
> 
> @zhyy2008z if due to the GFW，i mean，the inner ILSpy should also works as well，but it seems not work and fallback to the view of metadata

Without GFW, you can use the source link, that's what I found! Please forgive me for not mentioning inner ILSpy, I haven't looked into it.

**Comment by KieranDevvs** [comment]
> Whats GFW?

With the users appearing to be from Asia, I my guess would be GFW stands for "The Great Firewall". Basically China blocking western servers?

**Comment by cjmld5** [comment]
> Whats GFW?
> 
> The reason for the lack of real source in your situation is likely due to targeting .NET 6, but using a later SDK. In that case your projects are probably referencing nuget packages that literally do not contain any code, or SourceLink information, so there is nothing that Roslyn can use to find the right code. At runtime you'll roll forward and use whatever SDK is installed based on various rules and logic in the SDK resolvers, but Roslyn does not have that logic or information. We made it a little smarter, and it has some logic to pull the real info from SDK targetting packs, and nuget packages that contain both reference and implementation assemblies, but it doesn't cover all scenarios.
> 
> There is probably a little bit of input in the Output window, the "Navigate to External Sources" category, that could validate my guess here.

i see then output window，and sometime it says
1:  timeout  of downloading from pdb server,sometime 
2:  timeout of downloading from SourceLink，
3:  cant find pdb from sysbol server
4:  cant find document infomation in the pdb
and each case can not fallback to the inner ILSpy
i think should we could put inner ILSpy before metaview? or in some error like network errors，ILSPY can decompile it ,and the metaview to be the laststep？

**Comment by zhyy2008z** [comment]
VS 2022 17.7.3 built-in ilspy's version is too old, about two years ago. May be it cann't work well with reference assemblies?

![image](https://github.com/dotnet/roslyn/assets/5698463/a76d4171-65d1-45fa-a151-452e3c4c9f5b)

**Comment by CyrusNajmabadi** [comment]
We updated to 8.1 here: https://github.com/dotnet/roslyn/pull/69772/files

**Comment by davidwengier** [comment]
Reference assemblies don't have IL that can be decompiled into useful code. If the downloads timeout, we fall back to decompilatuon if we can, and only if that fails do we use metadata. Also if a download timeout, we stop making you wait, but continue the download in the background, so a future attempt to go to something from that same assembly could succeed.

**Comment by dgrunwald** [comment]
It's Roslyn that provides the list of assemblies that the decompiler should load; the decompiler engine doesn't pick files by itself.
If VS ends up uselessly decompiling references assemblies, that's a problem in Roslyn.

**Comment by CyrusNajmabadi** [comment]
@dgrunwald prs welcome :-)

**Comment by KieranDevvs** [comment]
> @dgrunwald prs welcome :-)

While I get the point you're making, you've posted this twice now and it's quite condescending to keep posting this any time anyone has criticism.

**Comment by davidwengier** [comment]
> If VS ends up uselessly decompiling references assemblies, that's a problem in Roslyn.

To be very clear here: The decompiler is doing nothing wrong, and VS doesn't "end up uselessly decompiling references assemblies". Roslyn is smart enough to know if something is a reference assembly, and in those cases we don't bother calling the decompiler. ILSpy is a wonderful piece of software.

The problem here, as I've written in many comments in this thread, is that we don't have a way of finding the implementation assemblies, and in some cases its entirely possible that the implementation assemblies do not exist on the users machine at all. From a Roslyn point of view as the code is currently written, if a user asks to Go To Definition for a method in .NET 6, then that is what we will try to do. Rolling forward to .NET 8 might happen at runtime on that machine, but its not the code being referenced by that project, so we won't show it. Relaxing that rule, and writing the logic for how to do that lookup, is what it would take to move this issue forward from the Roslyn side.

There are other approaches outside of Roslyn that could help here (see https://github.com/dotnet/sdk/issues/12360 for example) but none of it is straight forward.

**Comment by CyrusNajmabadi** [comment]
@zhyy2008z they likely have some heuristic to do the mapping.   As david mentioned, that's something that could be looked into.  And we'd probably be open to PRs that help out here, if they're suitably sized and reasonable to maintain.  Thanks!

**Comment by siegfriedpammer** [comment]
@zhyy2008z There's always the option to use the logic we employ in ILSpy to find implementation assemblies. However, this means that the framework(s) in question have to be installed on the machine and it would be necessary to implement some "ask the user which assembly to use" escape hatch. In ILSpy this is done by first looking in the active assembly list for references and using them. Only if an assembly cannot be resolved from the current list, we go looking for it in the filesystem using a myriad of hacks and crutches which luckily work most of the time.

@CyrusNajmabadi Do you think it's worth the hassle implementing this feature in VS? Ideally I would prefer to wait for an official solution from the people who designed .NET.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi Do you think it's worth the hassle implementing this feature using the code from ILSpy in VS? Ideally I would prefer to wait for an official solution from the people who designed .NET.

I'm fine with either.

**Comment by kirsan31** [comment]
Can anyone help with https://github.com/dotnet/roslyn/issues/68883#issuecomment-2404512342 🙏?

**Comment by xiangxingze** [comment]
Never found a good way in the VS good integrated source code view, ReSharper installed and uninstalled a very N times roslyn issue tracking, but also 6 years, and do not add this feature, I also want to traitor to Rider, our company's License to expire, to the point of not renewing it!

**Comment by CyrusNajmabadi** [comment]
@sharwell  I'm closing this out.  High level design discussion can continue in https://github.com/dotnet/roslyn/issues/75768.  Bugs can be opened as issues tracking particular problems.
